[Bug 1520918] [UI] Add UI for the denylist

OjaswinM opened the pull request:
Bugzilla Bug: [1520918](https://bugzilla.mozilla.org/show_bug.cgi?id=1520918)

Relates to #282

@helfi92, I made the initial table and the notifications are being listed correctly. Kindly check the progress till now and suggest changes.

To-dos:-

- Create one more view linked to the `+` icon which will be used to add addresses.
- Right now I'm not passing the description for the `HelpView` so I plan to create a file which uses the `RouteWithProps` component to either route to the list or add notifications view with the relevant props.
- Implement the delete functionality

helfi92 commented on ui/src/components/Dashboard/SidebarList.jsx:
```diff
@@ -13,6 +13,7 @@ import ScaleBalanceIcon from 'mdi-react/ScaleBalanceIcon';
 import ArrowExpandVerticalIcon from 'mdi-react/ArrowExpandVerticalIcon';
 import AccountSettingsIcon from 'mdi-react/AccountSettingsIcon';
 import GithubBoxIcon from 'mdi-react/GithubBoxIcon';
+import BellOffIcon from 'mdi-react/BellOffIcon';
```
Let's try using the `CancelIcon`.

helfi92 commented on ui/src/views/Denylist/index.jsx:
```diff
@@ -0,0 +1,56 @@
+    return (
+      <Dashboard
+        title="Notifications Denylist"
+        helpView={<HelpView description="description" />}
```
Let's make the description give information about the view.

helfi92 commented on ui/src/components/DenylistTable/index.jsx:
```diff
@@ -0,0 +1,140 @@
+                  <DeleteIcon size={iconSize} />
+                </Link>
+              </div>
+            </TableCell>
```
The view responsible for creation could also be responsible for deleting. This would remove the need of adding an extra column here. Similar to something like https://taskcluster-web.netlify.com/hooks/garbage/abc456.

OjaswinM replied:
How will we implement that? 

In [https://taskcluster-web.netlify.com/hooks](https://taskcluster-web.netlify.com/hooks) the delete button appears after we click a hook to view its details but we don't have a details page for the notification entries.

helfi92 replied:
In https://taskcluster-web.netlify.com/hooks, the delete page is the same page as the create page. The are both using the `HookForm` component. In our case, we could probably have a `DenyAddressForm` component? and show a delete button if it it does not have a `newDenyAddress` prop. Let me know if that doesn't make sense.

OjaswinM replied:
Ahh, okay I got what you mean :)

helfi92 commented on ui/src/utils/prop-types.js:
```diff
@@ -361,6 +361,11 @@ export const secrets = arrayOf(
 );
 
+export const notificationAddress = shape({
+  notificationType: oneOf(['EMAIL', 'PULSE', 'IRC_CHANNEL', 'IRC_USER']),
```
Create an entry for these constants in `src/utils/prop-types`.

helfi92 commented on ui/src/views/Denylist/index.jsx:
```diff
@@ -0,0 +1,56 @@
+        <Denylist searchTerm={notificationsSearch} />
+        <Button
+          spanProps={{ className: classes.plusIconSpan }}
+          tooltipProps={{ title: 'Add new address to denylist' }}
```
Make the tooltip shorter. Maybe "Add Address"?

OjaswinM commented:
@helfi92 I got the new views working :)
There are still minor bugs in the denylist which I'll fix in the next commit. I just wanted you to show the progress so you can review it.

Also I'm not sure about the taskcluster task failing message and the merge conflict.

OjaswinM commented:
> Add a mapping method that will help remove the capitals and underscores (i.e, "Email", "IRC User", ...)

Should I do this in only the drop down or the `Notification Type` column in the denylist as well?

> Navigating to the preview site of this PR https://deploy-preview-311--taskcluster-web.netlify.com/denylist shows:
> 
> <img alt="screen shot 2019-03-03 at 9 55 53 am" width="1665" src="https://user-images.githubusercontent.com/3766511/53697036-fbc8a780-3d9a-11e9-85df-ca7cc7e6e962.png">

I'm not sure why this is happening. The console shows that it is a graphql error but the query seems to be running fine on my side.

> What task failing message?

Nvm, that was due to a failing test. It's fixed now
 
> Look into doing a rebase. If you've never done this, you'll probably want to read about it first :)

Yeah that did it :)

djmitche commented:
@OjaswinM is this ready for review and merging?  @helfi92 is out for the week but I'll be happy to take a look when it's ready.

OjaswinM commented:
@djmitche, the `/denylist` view is running fine on my pc but shows an error here: [https://deploy-preview-311--taskcluster-web.netlify.com/denylist](https://deploy-preview-311--taskcluster-web.netlify.com/denylist). I'm not sure why

djmitche commented:
Hm, it looks like the `web-server` instance that is using needs to be deployed to add the endpoint.  I'll check into it.

OjaswinM commented:
@djmitche ahh okay.

Also, is there any way for me to check the graphql mutations without the proper auth?

djmitche commented:
At the moment it's a little tricky (since the site doesn't support logins yet), so I'll check it for you.

OjaswinM commented:
@djmitche Okay, great! I'll try to wrap everything else up tonight

OjaswinM commented:
@djmitche So other than checking the mutations I feel it's done now. There's just a small problem:- 

Steps to reproduce:-
1. Change the `VIEW_DENYLISTED_NOTIFICATIONS_PAGE_SIZE` to `2` in `/ui/utils/constants.js`
2. navigate to localhost:5080/denylist (it should show 2 entries per page).
3. now search `sher` in the search bar. This should match 2 entries in the current list(the first matched address is on page 1 and the second on page 2)

Expected:- 
After searching, both entries load on the same page

Actual:-
It shows one entry per page hence loading 2 pages for the 2 matched entries

djmitche commented:
That behavior is expected -- Azure does the pagination on the backend, and then we filter on the frontend.  It's actually possible to get pages with zero results.

djmitche commented:
I've updated tc-web-server now, so the https://deploy-preview-311--taskcluster-web.netlify.com/denylist link works again :)

djmitche commented on ui/src/components/DenylistForm/index.jsx:
```diff
@@ -0,0 +1,192 @@
+            <ListItem>
+              <ListItemText
+                primary="Notification Type"
+                secondary={notificationType}
```
prettify(..)

djmitche commented on ui/src/components/DenylistTable/index.jsx:
```diff
@@ -0,0 +1,140 @@
+          <TableRow key={node.notificationAddress}>
+            <TableCell>
+              <div>
+                <Typography>{node.notificationType}</Typography>
```
prettify(..)

djmitche commented on ui/src/components/DenylistTable/index.jsx:
```diff
@@ -0,0 +1,140 @@
+                className={classes.tableCell}
+                to={`/denylist/${encodeURIComponent(
+                  node.notificationAddress
+                )}`}>
```
Is it possible for the whole row to be a link?

helfi92 replied:
We should aim to be consistent across the UI. For tables with multiple columns, we don't typically have the whole row become a link. I see two options:

1. Have the hover effect on the first column only.
2. Remove the second column and have the whole row be a link. Only do this if the second column is not critical when listing the addresses.

OjaswinM replied:
I feel we should go with 1 since say, in the future we may have (in an unlike scenario) an irc user and an irc channel with very similar or same names which might cause confusion. Also, sorting would be easier.

@helfi92 @djmitche thoughts? :)

helfi92 replied:
Option 1 sounds good to me.

djmitche replied:
Same, and good point about consistency.

OjaswinM commented:
Ohh awesome! If the pagination is fine then  only thing left is checking the mutations now, i guess

OjaswinM commented on ui/src/components/Dashboard/SidebarList.jsx:
```diff
@@ -60,6 +61,9 @@ export default class SidebarList extends Component {
           </SidebarListItem>
+          <SidebarListItem to="/denylist" icon={<CancelIcon />}>
+            Denylist
+          </SidebarListItem>
```
Actually @helfi92 suggested that we should put it here

djmitche replied:
Hm, OK -- let's wait until he's back to figure that out then.

djmitche replied:
Let's call it "Notifications Denylist" then.  It's specific to notifications, so I don't want the UI to suggest that it somehow applies to all of Taskcluster.

OjaswinM commented on ui/src/components/DenylistForm/index.jsx:
```diff
@@ -0,0 +1,192 @@
+    notificationType: '',
+    notificationAddress: '',
+    // eslint-disable-next-line react/no-unused-state
+    prevAddress: null,
```
Ahh right, I need to update the prevAddress in `getDerivedStateFromProps`

OjaswinM commented on ui/src/components/DenylistForm/index.jsx:
```diff
@@ -0,0 +1,192 @@
+    };
+  }
+
+  static propTypes = {
```
Actually, I referenced `components/ClientForm` and `component/RoleForm` for this and both declare it after `defaultProps`

helfi92 replied:
I agree with @djmitche. Having both `propTypes` and `defaultProps` at the top helps in terms of readability. Feel free to do the change to `components/ClientForm` and `component/RoleForm`.

OjaswinM commented on ui/src/components/DenylistForm/index.jsx:
```diff
@@ -0,0 +1,192 @@
+                onChange={this.handleInputChange}
+                name="notificationType">
+                <MenuItem value="">
+                  <em>None</em>
```
Should I replace it with something like "Select type" ?

djmitche replied:
Or maybe simpler to just default to the first one?

OjaswinM commented:
@djmitche about making the whole row clickable in `components/DenylistTable`, I made both the columns link to the notification, but I'm sure there's a better way to do it than this. I also tried moving `<Link>` out so it encompasses `<TableRow>` but that messes up the layout and I'm not sure why. I'll continue looking into it. Other than that, I feel I've made the requested changes. Kindly review :)

djmitche commented:
Sounds good.  Those sound like questions for @helfi92 -- I'm just keeping his seat warm for the week.  So let's see what he says next week.

helfi92 commented on ui/src/utils/prop-types.js:
```diff
@@ -359,6 +359,18 @@ export const secrets = arrayOf(
+  'PULSE',
+  'IRC_CHANNEL',
+  'IRC_USER',
+]);
```
In `src/utils/contants.js` create a variable `DENYLIST_NOTIFICATION_TYPES` and refer to it instead of hard-coding the values here.

```jsx
export const DENYLIST_NOTIFICATION_TYPES = {
  'EMAIL': 'EMAIL',
  // ...
};
```

helfi92 commented on ui/src/views/Denylist/ViewDenylistAddress/index.jsx:
```diff
@@ -0,0 +1,108 @@
+
+      this.setState({ error: null, loading: false });
+
+      const url = `/notify/denylist/${encodeURIComponent(notificationAddress)}`;
```
No need to store the `url` in a temporary variable since it's not used in one place.

OjaswinM commented on ui/src/components/DenylistTable/index.jsx:
```diff
@@ -0,0 +1,158 @@
+  rSort((a, b) => sort(a.node.notificationAddress, b.node.notificationAddress)),
+  map(({ node: { notificationAddress } }) => notificationAddress)
+);
+const tableHeaders = ['Notification Type', 'Notification Address'];
```
The address can be an email address, pulse routing key or an IRC user or channel name. Maybe we can rename them to `Type` and `Address` since, as you said, the dashboard title already indicates the purpose of the table

helfi92 replied:
Good plan.

OjaswinM commented on ui/src/views/Denylist/routes.js:
```diff
@@ -0,0 +1,26 @@
+    component: ViewDenylist,
+    path,
+    description:
+      'Manage the notifications denylist. This page allows you to view, modify or delete the denylisted addresses.',
```
Oh right, we can't :p 
I'll change that

OjaswinM commented:
Thanks for the review @helfi92 @djmitche 
I have college tests going on so might not be able to work on this for a couple of days. Still I'll try to complete this asap :)

OjaswinM commented:
Hey @helfi92, sorry for the delay :)
Kindly review.

helfi92 commented on ui/src/views/Denylist/ViewDenylist/index.jsx:
```diff
@@ -0,0 +1,56 @@
+            placeholder="Notification address contains"
+          />
+        }>
+        <Denylist searchTerm={notificationsSearch} />
```
We don't need to create this component. We could have it here.

OjaswinM replied:
@helfi92 We need to run the graphql query on every `searchTerm` state change that's why I made a separate component. I feel it's done similarly in `views/Roles/ViewRoles/`. On the other hand, components like `view/Scopes/ListScopes` don't do this because, instead of querying the server with the filter, they pass the `searchTerm` to the rendering component which then filters it. I feel passing the `searchTerm` as props is the easiest way to re execute the graphql query. What are your thoughts on this?

helfi92 replied:
Makes sense :)

helfi92 commented on ui/src/components/DenylistForm/index.jsx:
```diff
@@ -0,0 +1,238 @@
+    /** Set to `true` when adding a new address. */
+    isNewaddress: bool,
+    /** Callback function fired when an address is created. */
+    onAddAddress: func.isRequired,
```
Rename to `onAddressAdd`.

helfi92 commented on ui/src/components/DenylistForm/index.jsx:
```diff
@@ -0,0 +1,238 @@
+    /** Callback function fired when an address is created. */
+    onAddAddress: func.isRequired,
+    /** Callback function fired when an address is deleted. */
+    onDeleteAddress: func,
```
Rename to `onAddressDelete`.

OjaswinM commented:
@helfi92 I've made the requested changes. Kindly review :)

helfi92 commented on ui/src/components/DenylistForm/index.jsx:
```diff
@@ -0,0 +1,237 @@
+  handleInputChange = ({ target: { name, value } }) => {
+    this.setState({ [name]: value });
+
+    if (name === 'notificationType' && value !== 'EMAIL') {
```
Extract the email from `DENYLIST_NOTIFICATION_TYPES`.

helfi92 commented on ui/src/components/DenylistForm/index.jsx:
```diff
@@ -0,0 +1,237 @@
+  render() {
+    const { address, classes, isNewAddress, loading } = this.props;
+    const { notificationType, notificationAddress, validation } = this.state;
+    const notificationTypes = ['EMAIL', 'PULSE', 'IRC_USER', 'IRC_CHANNEL'];
```
Is this a duplicate of `DENYLIST_NOTIFICATION_TYPES`?

helfi92 commented on ui/src/components/DenylistTable/index.jsx:
```diff
@@ -0,0 +1,168 @@
+    ...theme.mixins.hover,
+  },
+}))
+export default class DenylistTable extends Component {
```
The first column is missing the link icon:

![Screen Shot 2019-03-19 at 5 51 15 PM](https://user-images.githubusercontent.com/3766511/54644954-0bb9da00-4a71-11e9-851a-b870aaf21b52.png)

OjaswinM commented:
@helfi92 I changed `DENYLIST_NOTIFICATION_TYPES` to an object and made the other requested changes. Kindly check if everything is working.

One more thing, in last commit, we renamed some variables in `RolesForm`. Although I tried to change it across all the files, do check the mutations related to `RolesForm` once to make sure I didn't break anything :p

helfi92 commented on ui/src/components/DenylistForm/index.jsx:
```diff
@@ -169,7 +169,7 @@ export default class DenylistForm extends Component {
                 onChange={this.handleInputChange}
                 name="notificationType">
-                {notificationTypes.map(type => (
+                {Object.keys(DENYLIST_NOTIFICATION_TYPES).map(type => (
```
Did you mean to do `Object.values` instead?

OjaswinM replied:
They are the same in out case but yes `Object.values` is more appropriate

helfi92 commented on ui/src/utils/prop-types.js:
```diff
@@ -360,7 +360,7 @@ export const secrets = arrayOf(
 );
 
-export const notificationType = oneOf(DENYLIST_NOTIFICATION_TYPES);
+export const notificationType = oneOf(Object.keys(DENYLIST_NOTIFICATION_TYPES));
```
`Object.values` instead of `Object.keys`?

OjaswinM commented on ui/src/components/DenylistForm/index.jsx:
```diff
@@ -98,7 +99,7 @@ export default class DenylistForm extends Component {
     this.setState({ [name]: value });
 
-    if (name === 'notificationType' && value !== 'EMAIL') {
+    if (name === DENYLIST_NOTIFICATION_TYPES.EMAIL && value !== 'EMAIL') {
```
Ahh my bad, I'll fix this

helfi92 commented on ui/src/components/DenylistForm/index.jsx:
```diff
@@ -0,0 +1,240 @@
+
+        return pretty;
+      })
+      .join(' ');
```
You can make use of https://github.com/blakeembrey/change-case#readme to do things like remove underscores.

helfi92 commented on ui/src/components/DenylistTable/index.jsx:
```diff
@@ -0,0 +1,171 @@
+    padding: theme.spacing.unit,
+  },
+  listLinkCell: {
+    ...theme.mixins.hover,
```
You'll want to add `...theme.mixins.listItemButton,` in addition so that the link icon is faded when doesn't have focus or hover.

helfi92 commented on ui/src/views/Denylist/ViewDenylistAddress/index.jsx:
```diff
@@ -0,0 +1,109 @@
+@hot(module)
+@withApollo
+@graphql(denylistAddressQuery, {
+  skip: ({ match: { params } }) => !params.notificationAddress,
```
This view will only render when there is a notification address so this check is probably not necessary. I think it's safe to remove the `skip` field.

OjaswinM replied:
I think we use the same view for adding new addresses in which case we need to skip the query

helfi92 commented on ui/src/views/Denylist/ViewDenylistAddress/index.jsx:
```diff
@@ -0,0 +1,109 @@
+    error: null,
+  };
+
+  handleDeleteAddress = async (notificationType, notificationAddress) => {
```
`handleAdressDelete`.

helfi92 commented on ui/src/views/Denylist/ViewDenylistAddress/index.jsx:
```diff
@@ -0,0 +1,109 @@
+    }
+  };
+
+  handleAddAddress = async (notificationType, notificationAddress) => {
```
`handleAddressAdd`.

helfi92 commented on ui/src/views/Denylist/ViewDenylistAddresses/index.jsx:
```diff
@@ -0,0 +1,56 @@
+    this.setState({ searchTerm });
+  };
+
+  handleCreate = () => {
```
What are we creating? Change to `handle<the-thing-that-we-are-creating>Create`.

OjaswinM replied:
This is to navigate to the Add Address page. I think the name should be `handleAddressAdd` but we have an identically named function in another file that handles the actual mutations to add address. Is it okay if we use the same name here or should go for something different?

helfi92 replied:
Feel free to use the same name.

OjaswinM commented:
@helfi92 kindly review :)

helfi92 commented on ui/package-lock.json:
```diff
@@ -0,0 +1,18726 @@
+{
```
Remove `package-lock.json`. We already use `yarn.lock` to lock the dependencies.

```bash
# installs dependencies
yarn install

# add a dependency
yarn add <package-name>

# remove a dependency
yarn remove <package-name>
```

helfi92 commented on ui/src/components/DenylistForm/index.jsx:
```diff
@@ -44,7 +45,7 @@ export default class DenylistForm extends Component {
     onAddressAdd: func.isRequired,
     /** Callback function fired when an address is deleted. */
-    onAddressDelete: func,
+    onAddressDelete: func.isRequired,
```
Remove `onAddressDelete` from `static defaultProps` since it is set as required.

helfi92 commented on ui/src/components/DenylistForm/index.jsx:
```diff
@@ -0,0 +1,239 @@
+    isNewAddress: false,
+    address: null,
+    loading: false,
+    onAddressDelete: null,
```
Remove `onAddressDelete` from here since it is required.

djmitche commented:
@allcontributors[bot] please add @OjaswinM for code!

helfi92 commented:
Filed https://bugzilla.mozilla.org/show_bug.cgi?id=1540734 re: Taskcluster bot issue.

OjaswinM commented:
@helfi92 Thank you for helping me with this :)

helfi92 commented:
@oksmelnik My pleasure. It was great working with you :)
